### ISSUE TITLE
analyzer sometimes fails to report ambiguous_import

### ISSUE THREAD

**Issue by srawlins** [dart-lang/linter#2470]
Analyzer has a diagnostic for when two imports provide an element of the same name:

```dart
// a.dart ////////////////////
class File {}

// b.dart ////////////////////
class File {}

// c.dart ////////////////////
import 'a.dart';
import 'b.dart';

main() {
  File(); // error: ambiguous_import
}
```

But this error does not appear in all cases. Here is an example using dart:io:

```dart
// a.dart ////////////////////
class File {}

// c.dart ////////////////////
import 'a.dart';
import 'dart:io';

main() {
  File(); // MISSING ERROR
}
```

I'm not sure if the `dart:` part is the key.

Flutter is hitting this in a few places, like https://github.com/flutter/flutter/blob/master/packages/flutter_goldens/test/flutter_goldens_test.dart with "package:file/file.dart" and "dart:io".

**Comment by bwilkerson** [comment]
The exception was added so that the core libraries could be extended without breaking previously existing code.

Given that, why do you think users would want this lint? What broken code will this help to catch?

**Comment by srawlins** [comment]
I think most users are unaware of this singular exception to library element name collisions. There are libraries in flutter's codebase which import both `dart:io` and `package:file/file.dart`, and reference `File`. I believe most developers wouldn't know which `File` is being used in such a library.

If I had files like that in my code, I would want a lint to let me know, so that I could hide names that are being shadowed.

**Comment by mateusfccp** [comment]
I agree with this. It's unexpected for me that it's not an error, but after the explanation of why this is this way, I can understand why.

Even so, having an opt-in would be interesting, as, most of the time, shadowing a core identifier is done without intention, and may be overlooked.
